v2/contribute/code-style/react.mdx

```diff
@@ -6,11 +6,46 @@ hide_title: true
 
 # React
 
-- For really complicated components (that have a complex state, or lots of internal functions), prefer to use class type components as opposed to functional types.
+## General advice 
 
+- Remember to add the JSX pragma and import
+- Prefer adding components as `const X: React.FC<PropType> = (props) => {...}` instead of a function returning `JSX.Element`
 - When using `useEffect`:
-    - If it depends on certain variables instide an object, do not make that object a dependency in the array. Instead, extract the variable from the object before the use effect, and put that variable in the dependency array.
+    - If it depends on certain variables inside an object, do not make that object a dependency in the array. Instead, extract the variable from the object before the use effect, and put that variable in the dependency array.
     - If you want it to run it just once (similar to `componentDidMount`), provide it an empty dependency array.
-    - If applicable, make sure to return a function that runs in case the dependency array has changed (or component is unmounting). In this function, you can tell the `useEffect` logic to not change state for example (in case there is an `async / await` operation happening in it.)
+    - If applicable, make sure to return a function that runs in case the dependency array has changed (or the component is unmounting). In this function, you can tell the `useEffect` logic to not change state for example (in case there is an `async / await` operation happening in it.)
 
-- Refreaining returning a component from a function. Instead, make that function itself a functional component. This will prevent that component from unmounting and mounting during rerenders. If you must return a component from a function, then the consumer of that function should use `useRef`.
\ No newline at end of file
+- Refrain from returning a component from a function. Instead, make that function itself a functional component. This will prevent that component from unmounting and mounting during rerenders. If you must return a component from a function, then the consumer of that function should use `useRef` or `useMemo`.
+
+## Feature and Theme components
+
+Complex functionality should be implemented in feature components and theme components should only have local, simple, and strictly UI-related states. Some examples:
+
+### Feature component state:
+- Information that is created in one step of the process and used in another. I.e., `loginAttemptInfo` in passwordless sign in/up
+- General error displayed in the sign-in screen but set in the form submit method
+
+### Theme component state:
+- Current value of each form field
```

Comment:
One could even have the form fields as part of the feature component state? Why won't we add this as a state in there?

Reply:
The list of form fields (labels, validators, etc.) comes from the feature component (it would usually be part of useChildProps).
We could also elevate the form state (current value, validation status, etc) up to the feature component level, but it's not necessary: 
- we don't use them outside of the `FormBase` component
- we don't want to control or update them from the outside
- they are "temporary": we don't want to persist/restore them after refreshes or navigation.